### ISSUE TITLE
zfs option request

### ISSUE THREAD

**Issue by tonecaster** [kkrruumm/void-install-script#6]
Thank you for your great script!
Is there any chance you could add ZFS option to the install?
I believe this could make you a true star in the linux, void-linux community. 
I've installed mine from a script I put together after many days of trials and errors and have been running this amazing very stable system for the last 2 years... 
My aim is to install an encrypted ZFS rooted void xfce with all its goodies and i3 as its WM on a new machine I'm buying with an amazing WYSIWYG install script like yours, if possible.

Thanks for the attention!

**Comment by kkrruumm** [comment]
ZFS is definitely planned at some point, though this installer has taken a bit of a backseat in favor of other projects I am currently part of unfortunately.

The best way to have ZFS here is to open a pull request adding what you would like, after review, I'm open to merging from others.

I think the best way to go about ZFS support would be zfsbootmenu as the only supported option, which will allow you to use ZFS native encryption rather than luks, in which case the installer should be ran from the hrmpf image here: [https://github.com/leahneukirchen/hrmpf](url)

I will look into it, but I can't tell you when this will be completed.

**Comment by tonecaster** [comment]
Thanks for your response!
As I've mentioned, I've been running void zfs on my system with the script I put up together and the hrmpf way is the road I took at the time... But I've always wanted a graphic installer like yours, or like the now defunct Trident Linux for that and it'd be great to have it...

Once again thanks!

**Comment by zenodotus280** [comment]
I created a ZFS-BM install script that follows the official instructions as closely as possible. I'll make it available soon and perhaps we can dove-tail them together?

**Comment by kkrruumm** [comment]
@zenodotus280 that sounds good, what i would appreciate the most is a pull request, i'm having issues finding time to add major features like that, i just got around to adding UKI support despite having daily driven it for months-

if you'd like to create a pull request, i'm more than willing to look over it and merge.

otherwise, after making it public, i'll do what i can to get this feature added, i hadn't meant to leave this issue open for so long :p

**Comment by zenodotus280** [comment]
Okay - mine is only about 300 lines long so I can probably strip out the ZBM-specific commands and fit them into yours. Probably won't get around to this for a few weeks or more... you know how it goes ;)

**Comment by kkrruumm** [comment]
that sounds pretty clean, but we also need to figure out what we're going to do with encryption

i've never tried ZBM, but if we can use luks instead of zfs native encryption i would prefer that, i don't have much faith in zfs native encryption at the moment after having looked into it for the past few months

let me know once yours goes public and i'll take a look around

**Comment by tonecaster** [comment]
Hi guys!
This is sounding good!
I've been using zfs native encryption on my void linux for more than 2 years now without a glitch. 
Very stable, secure and fast!
Can't recommend it enough...
Thanks for the time and effort you put on this...

**Comment by kkrruumm** [comment]
i'm less concerned with stability (ZFS has proven itself plenty stable) and more concerned with how zfs with encryption leaks metadata, there's no way to detach headers of course, and the number of issues that are still open related to specifically encryption

do check the "Caveats" part of this wiki page: https://wiki.gentoo.org/wiki/ZFS

i'm not entirely opposed to using zfs native encryption, just need to make sure this is handled correctly

as for boot chain security, it seems like ZFSBM is more or less just a UKI with a TUI and tools to handle ZFS, i like this a lot
i run secure boot with my own setups, protecting just the UKI, that being the only file actually exposed unencrypted on my disk so i'd like to take into account similar setups with ZFSBM if possible

**Comment by trolljoe** [comment]
>but if we can use LUKS instead of native encryption

you can ([alpine wiki](https://wiki.alpinelinux.org/wiki/Setting_up_ZFS_on_LUKS)), but i honestly think adding support for both would be nice.

**Comment by kkrruumm** [comment]
> you can ([alpine wiki](https://wiki.alpinelinux.org/wiki/Setting_up_ZFS_on_LUKS)), but i honestly think adding support for both would be nice.

support for both is valid, right now i'm working on a branch that makes LVM optional, zfs support will have to come after i push that to main

i'm also not a zfs user, so this is going to need some testing from people who do use it, especially as far as the settings deployed by default go

**Comment by kkrruumm** [comment]
lvm is now optional, revisiting this

**Comment by kkrruumm** [comment]
so as things stand at the moment, zfsbootmenu doesn't have a feature to unlock a partition that is encrypted with luks as opposed to zfs native encryption- i'm looking into writing a hook for zbm to do this, as it's a *much* nicer solution compared to pretty much anything else for this usecase

until then, zfs with zbm and zfs native encryption shouldn't be too much work to implement

**Comment by zenodotus280** [comment]
Can you direct me to some reading on how "ZFS leaks metadata"? Is this a real problem or just one of aesthetics/taste? Asking out of genuine curiosity.

**Comment by kkrruumm** [comment]
> Can you direct me to some reading on how "ZFS leaks metadata"? Is this a real problem or just one of aesthetics/taste? Asking out of genuine curiosity.

the metadata leaks aren't the end of the world- here's a little bit more information for you:
https://forums.truenas.com/t/truenas-zfs-encryption-deduplication-for-home-server/13589/3

my primary concern is just with having both as an option, but i hate to have to ship a third party hook along with zfsbootmenu for the time being to make both realistic

**Comment by zenodotus280** [comment]
Great, thanks. I had no idea! I don't think it matters much to me but _knowing_ that my pools, datasets, and snapshot names are in the open is important. I also understand your desire to just pick a single encryption method that protects everything rather than offer one that is tightly integrated but with caveats.

When you're looking for some testing to be done I'll give it a shot with a variety of configurations.

**Comment by kkrruumm** [comment]
yeah, and that's kind of the thing here- users need to be aware that this information is going to be visible, but in terms of the actual data contents, of course it's pretty much as good as gone if someone is getting their hands on your drive while it's cold

when support is first introduced i'm going to leave it limited to just zfs native encryption until i work out some hooks for zfsbootmenu to use luks instead, if the user would rather use that

>When you're looking for some testing to be done I'll give it a shot with a variety of configurations.

i'll send this over to a branch when it's ready to be tested, i appreciate it.

**Comment by kkrruumm** [comment]
@zenodotus280 @tonecaster @trolljoe 

the zfs branch is now up with base albeit fairly complete zfs support via zfsbootmenu: https://github.com/kkrruumm/void-install-script/tree/zfs

some things to note:
- the minimum passphrase length for encryption is 8 characters, the installer will boot you out if you do not meet that at the moment

- this is not ready to actually use in the context of encryption as i've still not decided how i want to go about calculating iterations to bump from the default of 350000

- i also added a setKernelParam wrapper function to get rid of the sed spam to adjust that

- there's the potential that i have broken things unrelated to zfs in the process of doing this

- zfs native encryption is the only supported option at the moment, as is zfsbootmenu

if any of you are still up for testing, let me know what you find (or don't find), this is likely to stay on its own branch for a few days until it's had time to marinate

**Comment by kkrruumm** [comment]
so, one thing- swapping on zfs is busted

at the moment, swapping on a zvol is still a bad idea because you're likely to deadlock the system under memory pressure, and normal swapfiles are a no-go as well

the sane routes here are to swap outside of zfs, such as creating a separate partition for swap which kinda sucks in the event you need to grow it, and it would need to be encrypted with luks and likely automatically unlocked via a luks key (similar to how FDE is handled with GRUB) **or** just use zram.

**Comment by zenodotus280** [comment]
I use ZRAM on my Void setup too.
I'll try this out next weekend and see how it is.

**Comment by kkrruumm** [comment]
finally got some time, so:

the installer will now calculate iterations for zfs' pbkdf2 dynamically via a ballpark calculation with cryptsetups benchmark, this isn't perfectly accurate as one would expect, so i've got it leaning into overshooting by about half a second on my system- a potential concern is this time variance changing per system, so this needs more testing but that was the final real "roadblock" preventing this from being merged, but it needs to be made reasonably certain that this is going to match the specified itertime (by default, 10 seconds) or overshoot it slightly as opposed to not meeting it.
